applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_gradient_projection.py

```diff
@@ -184,71 +148,73 @@ def __analyzeShape(self):
+            for constraint in self.constraints:
+                con_id = constraint["identifier"].GetString()
+                gradient_variable = self.constraint_gradient_variables[con_id]["gradient"]
+                mapped_gradient_variable = self.constraint_gradient_variables[con_id]["mapped_gradient"]
+                mapper.InverseMap(gradient_variable, mapped_gradient_variable)
 
-        self.__computeControlPointUpdate()
+            self.__computeControlPointUpdate()
 
-        self.mapper.Map(KSO.CONTROL_POINT_UPDATE, KSO.SHAPE_UPDATE)
-        self.model_part_controller.DampNodalVariableIfSpecified(KSO.SHAPE_UPDATE)
+            mapper.Map(KSO.CONTROL_POINT_UPDATE, KSO.SHAPE_UPDATE)
+            self.model_part_controller.DampNodalVariableIfSpecified(KSO.SHAPE_UPDATE)
 
     # --------------------------------------------------------------------------
     def __computeControlPointUpdate(self):
         """adapted from https://msulaiman.org/onewebmedia/GradProj_2.pdf"""
         g_a, g_a_variables = self.__getActiveConstraints()
 
-        KM.Logger.PrintInfo("ShapeOpt", "Assemble vector of objective gradient.")
-        nabla_f = KM.Vector()
-        s = KM.Vector()
-        self.optimization_utilities.AssembleVector(self.design_surface, nabla_f, KSO.DF1DX_MAPPED)
-
-        if len(g_a) == 0:
-            KM.Logger.PrintInfo("ShapeOpt", "No constraints active, use negative objective gradient as search direction.")
-            s = nabla_f * (-1.0)
-            s *= self.step_size / s.norm_inf()
-            self.optimization_utilities.AssignVectorToVariable(self.design_surface, s, KSO.SEARCH_DIRECTION)
-            self.optimization_utilities.AssignVectorToVariable(self.design_surface, [0.0]*len(s), KSO.CORRECTION)
-            self.optimization_utilities.AssignVectorToVariable(self.design_surface, s, KSO.CONTROL_POINT_UPDATE)
-            return
-
-
-        KM.Logger.PrintInfo("ShapeOpt", "Assemble matrix of constraint gradient.")
-        N = KM.Matrix()
-        self.optimization_utilities.AssembleMatrix(self.design_surface, N, g_a_variables)  # TODO check if gradients are 0.0! - in cpp
-
-        settings = KM.Parameters('{ "solver_type" : "LinearSolversApplication.dense_col_piv_householder_qr" }')
-        solver = dense_linear_solver_factory.ConstructSolver(settings)
-
-        KM.Logger.PrintInfo("ShapeOpt", "Calculate projected search direction and correction.")
-        c = KM.Vector()
-        self.optimization_utilities.CalculateProjectedSearchDirectionAndCorrection(
-            nabla_f,
-            N,
-            g_a,
-            solver,
-            s,
-            c)
-
-        if c.norm_inf() != 0.0:
-            if c.norm_inf() <= self.max_correction_share * self.step_size:
-                delta = self.step_size - c.norm_inf()
-                s *= delta/s.norm_inf()
+        for name, design_surface in self.design_surfaces.items():
```

Comment:
With such a loop over design surfaces at this stage, the optimization is separated into two independent optimizations, that are only connected due to the simultaneous evaluation of the responses. 

The problematic part is that the update of the design surfaces is computed individually instead of simultaneously. This is less relevant for a steepest descent algorithm, but can change the path to optimum for constrained cases quite drastically.

I would actually prefer if all sensitivities from all design surfaces are assembled into one global vector and the update is only computed once, since that seems the more natural approach to me.

Interested to hear @dbaumgaertner  opinion on this as well. 

@adityaghantasala Do you have a special requirement to deal with the individual design surfaces independenty?

Reply:
I see your point. I will see if I can do gradient projection for all the design surfaces. 

About requirement to deal with individual design surfaces independently, sort of yes. The design surface I have will require different filter radii as they have features of different sizes. And many times these surfaces are completely disjoint and way apart. 
On a whole, having the ability to deal with multiple design surface is important, if not immediately, but definitely in long run.

Reply:
>  The design surface I have will require different filter radii as they have features of different sizes. And many times these surfaces are completely disjoint and way apart.

This i see, the main question is if a sequential optimization (ev. problematic for constrained problems) or a simultaneous optimization (ev. challenging due to different scale of the variables) is more appropriate for the problem. 

I have the feeling that the simultaneous approach is something that has to be implemented in the algorithm level (ev. also with a special algorithm that inherits from the standard algorithm with just one design surface).

Having the algoritms low level functions working on assembled vectors and matrices (as in GradientProjection.py) would allow them to be easily reused for both scenarios.

The sequential approach could ev. also be implemented like coupled analyses: For example with a function like `PerformOptimizationStep` and then doing the main loop of two distinct optimization objects somewhere in a wrapping object or custom code in a runscript.

Reply:
> I have the feeling that the single/simultaneous approach is something that has to be implemented in the algorithm level (ev. also with a special algorithm that inherits from the standard algorithm with just one design surface).
> Having the algoritms low level functions working on assembled vectors and matrices (as in GradientProjection.py) would allow them to be easily reused for both scenarios.

I would say it should be other way around. that means, having the possibility to do single optimization should be in a derived class.

I am already close to finishing the implementation of simultaneous approach. I will push it today. We can discuss on that even more. 

One more thing that I observed for single/sequential optimization is that, the step size is applied to each design surface individually but not to the whole problem. This could be a problem and should be addressed.

Reply:
> One more thing that I observed for single/sequential optimization 

You mean single/simultaneous opt? Yes that has to be adressed, it is not allowed to apply different scaling to the update of the design surfaces after the update was computed.

Reply:
I have just pushed a working version of simultaneous update calculation in gradient projection. 
I assemble all the nabla_f, and N and calculate the s and c vectors for all the given design surfaces, do scaling and then set them to the respective design surfaces. 

do have a look.